File: leetcode_python/Tree/all-nodes-distance-k-in-binary-tree.py
"""

863. All Nodes Distance K in Binary Tree
Medium


Given the root of a binary tree, the value of a target node target, and an integer k, return an array of the values of all nodes that have a distance k from the target node.

You can return the answer in any order.

 

Example 1:


Input: root = [3,5,1,6,2,0,8,null,null,7,4], target = 5, k = 2
Output: [7,4,1]
Explanation: The nodes that are a distance 2 from the target node (with value 5) have values 7, 4, and 1.
Example 2:

Input: root = [1], target = 1, k = 3
Output: []
 

Constraints:

The number of nodes in the tree is in the range [1, 500].
0 <= Node.val <= 500
All the values Node.val are unique.
target is the value of one of the nodes in the tree.
0 <= k <= 1000

"""



# V0
# IDEA: DFS + BFS (GPT)
"""
CORE IDEA:


-> Use BFS to radiate outward from the target node


1. build the graph: {node: parent}
2. use BFS, init q as [target, 0]    (node, dist)
3. run BFS, collect nodes when dist == k
"""
from collections import defaultdict, deque

class Solution(object):
    def distanceK(self, root, target, k):
        self.graph = defaultdict(list)
        self.build_graph(root)

        res = []
        visited = {target}
        q = deque([(target, 0)])

        while q:
            node, dist = q.popleft()

            if dist == k:
                res.append(node.val)
                continue

            for nxt in self.graph[node]:
                if nxt not in visited:
                    visited.add(nxt)
                    q.append((nxt, dist + 1))

        return res


    # NOTE !!
    # helper func ONLY has 1 param
    def build_graph(self, root):
        if not root:
            return

        """
        NOTE !!!


        1. we need to record `bi - direction` path
           -> e.g.

             root <--> root.left



        2. ONLY do above if child exists
            e.g.

            if root.left
            if root.right

        """
        if root.left:
            self.graph[root].append(root.left)
            self.graph[root.left].append(root)

        if root.right:
            self.graph[root].append(root.right)
            self.graph[root.right].append(root)

        self.build_graph(root.left)
        self.build_graph(root.right)


# V0
# IDEA: DFS + BFS (gemini)
"""
CORE IDEA:


-> Use BFS to radiate outward from the target node


1. build the graph: {node: parent}
2. use BFS, init q as [target, 0]    (node, dist)
3. run BFS, collect nodes when dist == k
"""
from collections import defaultdict, deque

class Solution(object):
    def distanceK(self, root, target, k):
        self.graph = defaultdict(list)

        self.build_graph(root, None)

        q = deque([(target, 0)])

        # NOTE !!!
        # use `visited` to avoid duplicated visiting
        visited = {target}
        res = []

        while q:
            node, dist = q.popleft()

            if dist == k:
                res.append(node.val)
                continue

            for nxt in self.graph[node]:
                if nxt not in visited:
                    visited.add(nxt)
                    q.append((nxt, dist + 1))

        return res


    # pre-order DFS
    # NOTE !!
    # helper func has 2 param (e.g. node, parent)
    def build_graph(self, node, parent):
        if not node:
            return

        """
        NOTE !!!

        how we build the graph

        -> move `2 directions`
        -> e.g.  node -> parent, parent -> node
            -> so we can both record the relation from node <--> parent


        -> also NOTE the `if parent` condition
        """
        if parent:
            self.graph[node].append(parent)
            self.graph[parent].append(node)


        # NOTE: dfs calling
        self.build_graph(node.left, node)
        self.build_graph(node.right, node)


# V0-0-1
# IDEA: BFS (gemini)
"""
CORE IDEA:


-> Use BFS to radiate outward from the target node


1. build the graph: {node: parent}
2. use BFS, init q as [target, 0]    (node, dist)
3. run BFS, collect nodes when dist == k
"""
import collections

class Solution(object):
    def distanceK(self, root, target, k):
        """
        :type root: TreeNode
        :type target: TreeNode
        :type k: int
        :rtype: List[int]
        """
        # Step 1: Traverse the tree to map every node to its parent
        parents = {}
        def add_parents(node, parent):
            if not node:
                return
            parents[node] = parent
            add_parents(node.left, node)
            add_parents(node.right, node)
            
        add_parents(root, None)
        
        # Step 2: Use BFS to radiate outward from the target node
        queue = collections.deque([(target, 0)]) # (current_node, distance)
        visited = set([target])
        ans = []
        
        while queue:
            node, dist = queue.popleft()
            
            # If we reached distance k, add to answer. 
            # We don't need to go any further down this path!
            if dist == k:
                ans.append(node.val)
                continue
                
            # Look at all 3 possible directions: Left, Right, and Up (Parent)
            for neighbor in (node.left, node.right, parents[node]):
                # If the neighbor exists and we haven't visited it yet
                if neighbor and neighbor not in visited:
                    visited.add(neighbor)
                    queue.append((neighbor, dist + 1))
                    
        return ans


# V0-1

# V0-2

# V0-3
# IDEA: BFS (gemini)
"""
CORE IDEA:


-> Use BFS to radiate outward from the target node


1. build the graph: {node: parent}
2. use BFS, init q as [target, 0]    (node, dist)
3. run BFS, collect nodes when dist == k
"""
import collections

class Solution(object):
    def distanceK(self, root, target, k):
        """
        :type root: TreeNode
        :type target: TreeNode
        :type k: int
        :rtype: List[int]
        """
        # Step 1: Traverse the tree to map every node to its parent
        parents = {}
        def add_parents(node, parent):
            if not node:
                return
            parents[node] = parent
            add_parents(node.left, node)
            add_parents(node.right, node)
            
        add_parents(root, None)
        
        # Step 2: Use BFS to radiate outward from the target node
        queue = collections.deque([(target, 0)]) # (current_node, distance)
        visited = set([target])
        ans = []
        
        while queue:
            node, dist = queue.popleft()
            
            # If we reached distance k, add to answer. 
            # We don't need to go any further down this path!
            if dist == k:
                ans.append(node.val)
                continue
                
            # Look at all 3 possible directions: Left, Right, and Up (Parent)
            for neighbor in (node.left, node.right, parents[node]):
                # If the neighbor exists and we haven't visited it yet
                if neighbor and neighbor not in visited:
                    visited.add(neighbor)
                    queue.append((neighbor, dist + 1))
                    
        return ans

# V0-4
# IDEA: DFS (gpt)
class Solution(object):
    def distanceK(self, root, target, k):
        self.ans = []
        self.k = k

        self.dfs(root, target)
        return self.ans

    def collect(self, root, dist):
        if not root:
            return

        if dist == self.k:
            self.ans.append(root.val)
            return

        self.collect(root.left, dist + 1)
        self.collect(root.right, dist + 1)

    def dfs(self, root, target):
        if not root:
            return -1

        if root == target:
            self.collect(root, 0)
            return 0

        left = self.dfs(root.left, target)

        if left != -1:
            if left + 1 == self.k:
                self.ans.append(root.val)
            else:
                self.collect(root.right, left + 2)
            return left + 1

        right = self.dfs(root.right, target)

        if right != -1:
            if right + 1 == self.k:
                self.ans.append(root.val)
            else:
                self.collect(root.left, right + 2)
            return right + 1

        return -1


# V0
# IDEA : DFS + BFS
# https://leetcode.com/problems/all-nodes-distance-k-in-binary-tree/discuss/604718/Python-BFS-solution
# DEMO
# root = TreeNode{val: 3, left: TreeNode{val: 5, left: TreeNode{val: 6, left: None, right: None}, right: TreeNode{val: 2, left: TreeNode{val: 7, left: None, right: None}, right: TreeNode{val: 4, left: None, right: None}}}, right: TreeNode{val: 1, left: TreeNode{val: 0, left: None, right: None}, right: TreeNode{val: 8, left: None, right: None}}}
# self.graph = defaultdict(<class 'list'>, {3: [5, 1], 5: [3, 6, 2], 6: [5], 2: [5, 7, 4], 7: [2], 4: [2], 1: [3, 0, 8], 0: [1], 8: [1]})
# time = O(n)  # n = number of tree nodes
# space = O(n)
from collections import defaultdict
class Solution:
    
    def build(self,parent,child):
        if parent and child:
            self.graph[parent.val].append(child.val)
            self.graph[child.val].append(parent.val)
        if child.left:
            self.build(child,child.left)
        if child.right:
            self.build(child,child.right)
            
    def distanceK(self, root, target, K):
        self.graph=defaultdict(list)
        self.build(None,root)
        print ("root = " + str(root))
        print ("self.graph = " + str(self.graph))
        q=[(target.val,1)]
        vis=set([target.val])
        ans=[]
        while q:
            i,j=q.pop(0)
            for node in self.graph[i]:
                if node not in vis:
                    if j==K:
                        ans.append(node)
                    vis.add(node)
                    q.append((node,j+1))
        return ans if len(q) < K else [target.val]

# V0'
# IDEA : DFS + BFS
# time = O(n)  # n = number of tree nodes
# space = O(n)
class Solution(object):
    def distanceK(self, root, target, K):
        # DFS
        def connect(parent, child):
            # build graph
            if parent and child:
                conn[parent.val].append(child.val)
                conn[child.val].append(parent.val)
            if child.left:
                connect(child, child.left)
            if child.right:
                connect(child, child.right)
        # init
        conn = collections.defaultdict(list)
        connect(None, root)
        print ("root = " + str(root))
        print ("conn = " + str(conn))
        # BFS
        #q = collections.deque()
        """
        NOTE THIS !!!
        """
        q = [target.val]
        #q.append(target.val)
        """
        NOTE THIS !!!
        """
        visited = set([target.val])
        for k in range(K):
            size = len(q)
            for i in range(size):
                node = q.pop(0)
                for j in conn[node]:
                    if j not in visited:
                        q.append(j)
                        visited.add(j)
        return list(q)

# V1
# IDEA : DFS + BFS
# https://leetcode.com/problems/all-nodes-distance-k-in-binary-tree/discuss/604718/Python-BFS-solution
# time = O(n)  # n = number of tree nodes
# space = O(n)
from collections import defaultdict,deque
class Solution:
    def __init__(self):
        self.graph=defaultdict(list)
    
    def build(self,parent,child):
        if parent and child:
            self.graph[parent.val].append(child.val)
            self.graph[child.val].append(parent.val)
        if child.left:
            self.build(child,child.left)
        if child.right:
            self.build(child,child.right)
            
    def distanceK(self, root, target, K):
        self.build(None,root)
        q=deque()
        q.append((target.val,1))
        vis=set([target.val])
        ans=[]
        while q:
            i,j=q.popleft()
            for node in self.graph[i]:
                if node not in vis:
                    if j==K:
                        ans.append(node)
                    vis.add(node)
                    q.append((node,j+1))
        return ans if len(q) < K else [target.val]

# V1'
# IDEA :  Annotate Parent
# https://leetcode.com/problems/all-nodes-distance-k-in-binary-tree/solution/
# time = O(n)  # n = number of tree nodes
# space = O(n)
class Solution(object):
    def distanceK(self, root, target, K):
        def dfs(node, par = None):
            if node:
                node.par = par
                dfs(node.left, node)
                dfs(node.right, node)

        dfs(root)

        queue = collections.deque([(target, 0)])
        seen = {target}
        while queue:
            if queue[0][1] == K:
                return [node.val for node, d in queue]
            node, d = queue.popleft()
            for nei in (node.left, node.right, node.par):
                if nei and nei not in seen:
                    seen.add(nei)
                    queue.append((nei, d+1))

        return []

# V1''
# IDEA : Percolate Distance
# https://leetcode.com/problems/all-nodes-distance-k-in-binary-tree/solution/
# time = O(n)  # n = number of tree nodes
# space = O(n)
class Solution(object):
    def distanceK(self, root, target, K):
        ans = []

        # Return distance from node to target if exists, else -1
        # Vertex distance: the # of vertices on the path from node to target
        def dfs(node):
            if not node:
                return -1
            elif node is target:
                subtree_add(node, 0)
                return 1
            else:
                L, R = dfs(node.left), dfs(node.right)
                if L != -1:
                    if L == K: ans.append(node.val)
                    subtree_add(node.right, L + 1)
                    return L + 1
                elif R != -1:
                    if R == K: ans.append(node.val)
                    subtree_add(node.left, R + 1)
                    return R + 1
                else:
                    return -1

        # Add all nodes 'K - dist' from the node to answer.
        def subtree_add(node, dist):
            if not node:
                return
            elif dist == K:
                ans.append(node.val)
            else:
                subtree_add(node.left, dist + 1)
                subtree_add(node.right, dist + 1)

        dfs(root)
        return ans

# V1''' 
# https://blog.csdn.net/fuxuemingzhu/article/details/82709619
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
# time = O(n)  # n = number of tree nodes
# space = O(n)
class Solution(object):
    def distanceK(self, root, target, K):
        """
        :type root: TreeNode
        :type target: TreeNode
        :type K: int
        :rtype: List[int]
        """
        # DFS
        conn = collections.defaultdict(list)
        def connect(parent, child):
            if parent and child:
                conn[parent.val].append(child.val)
                conn[child.val].append(parent.val)
            if child.left: connect(child, child.left)
            if child.right: connect(child, child.right)
        connect(None, root)
        # BFS
        que = collections.deque()
        que.append(target.val)
        visited = set([target.val])
        for k in range(K):
            size = len(que)
            for i in range(size):
                node = que.popleft()
                for j in conn[node]:
                    if j not in visited:
                        que.append(j)
                        visited.add(j)
        return list(que)

### Test case : dev 

# V1''''
# https://leetcode.com/problems/all-nodes-distance-k-in-binary-tree/discuss/1044354/easy-python
# time = O(n)  # n = number of tree nodes
# space = O(n)
class Solution:
    def distanceK(self, root: TreeNode, target: TreeNode, K: int) -> List[int]:
        def dfs(node, dist):
            if not node:
                return
            
            cur_dist = dist
            if node.val in path:
                cur_dist = path[node.val]
                
            if cur_dist == K:
                res.append(node.val)
                
            dfs(node.left, cur_dist+1)
            dfs(node.right, cur_dist+1)
            return 
        
        def find_path(node, path):
            if not node:
                return -1
            
            if node.val == target.val:
                path[node.val] = 0
                return 0
            
            left = find_path(node.left, path)
            if left >= 0:
                path[node.val] = left+1
                return left+1
            
            right = find_path(node.right, path)
            if right >= 0:
                path[node.val] = right+1
                return right+1
            
            return -1
        
        res = []
        path = {}           
        find_path(root, path)
        dfs(root, path[root.val])
        return res

# V1''''''
# https://blog.csdn.net/fuxuemingzhu/article/details/82709619
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
# time = O(n)  # n = number of tree nodes
# space = O(n)
class Solution(object):
    def distanceK(self, root, target, K):
        """
        :type root: TreeNode
        :type target: TreeNode
        :type K: int
        :rtype: List[int]
        """
        # DFS
        conn = collections.defaultdict(list)
        def connect(parent, child):
            if parent and child:
                conn[parent.val].append(child.val)
                conn[child.val].append(parent.val)
            if child.left: connect(child, child.left)
            if child.right: connect(child, child.right)
        connect(None, root)
        # BFS
        bfs = [target.val]
        visited = set([target.val])
        for k in range(K):
            bfs = [y for x in bfs for y in conn[x] if y not in visited]
            visited |= set(bfs)
        return bfs

# V1'''''''
# https://www.jiuzhang.com/solution/all-nodes-distance-k-in-binary-tree/#tag-highlight-lang-python
# time = O(n)  # n = number of tree nodes
# space = O(n)
class Solution(object):
    def distanceK(self, root, target, K):
        def dfs(node, par = None):
            if node:
                node.par = par
                dfs(node.left, node)
                dfs(node.right, node)

        dfs(root)

        queue = collections.deque([(target, 0)])
        seen = {target}
        while queue:
            if queue[0][1] == K:
                return [node.val for node, d in queue]
            node, d = queue.popleft()
            for nei in (node.left, node.right, node.par):
                if nei and nei not in seen:
                    seen.add(nei)
                    queue.append((nei, d+1))
        return []

# V1''''''''''
# https://leetcode.com/problems/all-nodes-distance-k-in-binary-tree/solution/
# IDEA :  Annotate Parent
# PROCESS:
# 0. If we know the parent of every node x, we know all nodes that are distance 1 from x. We can then perform a breadth first search from the target node to find the answer.
# 1. We first do a depth first search where we annotate every node with information about it's parent.
# 2. After, we do a breadth first search to find all nodes a distance K from the target.
# time = O(n)  # n = number of tree nodes
# space = O(n)
class Solution(object):
    def distanceK(self, root, target, K):
        def dfs(node, par = None):
            if node:
                node.par = par
                dfs(node.left, node)
                dfs(node.right, node)

        dfs(root)

        queue = collections.deque([(target, 0)])
        seen = {target}
        while queue:
            if queue[0][1] == K:
                return [node.val for node, d in queue]
            node, d = queue.popleft()
            for nei in (node.left, node.right, node.par):
                if nei and nei not in seen:
                    seen.add(nei)
                    queue.append((nei, d+1))

        return []

# V1''''''''''
# https://leetcode.com/problems/all-nodes-distance-k-in-binary-tree/solution/
# IDEA : Percolate Distance
# time = O(n)  # n = number of tree nodes
# space = O(n)
class Solution(object):
    def distanceK(self, root, target, K):
        ans = []

        # Return distance from node to target if exists, else -1
        # Vertex distance: the # of vertices on the path from node to target
        def dfs(node):
            if not node:
                return -1
            elif node is target:
                subtree_add(node, 0)
                return 1
            else:
                L, R = dfs(node.left), dfs(node.right)
                if L != -1:
                    if L == K: ans.append(node.val)
                    subtree_add(node.right, L + 1)
                    return L + 1
                elif R != -1:
                    if R == K: ans.append(node.val)
                    subtree_add(node.left, R + 1)
                    return R + 1
                else:
                    return -1

        # Add all nodes 'K - dist' from the node to answer.
        def subtree_add(node, dist):
            if not node:
                return
            elif dist == K:
                ans.append(node.val)
            else:
                subtree_add(node.left, dist + 1)
                subtree_add(node.right, dist + 1)

        dfs(root)
        return ans

# V2
# time = O(n)
# space = O(n)
import collections
class Solution(object):
    def distanceK(self, root, target, K):
        """
        :type root: TreeNode
        :type target: TreeNode
        :type K: int
        :rtype: List[int]
        """
        def dfs(parent, child, neighbors):
            if not child:
                return
            if parent:
                neighbors[parent.val].append(child.val)
                neighbors[child.val].append(parent.val)
            dfs(child, child.left, neighbors)
            dfs(child, child.right, neighbors)

        neighbors = collections.defaultdict(list)
        dfs(None, root, neighbors)
        bfs = [target.val]
        lookup = set(bfs)
        for _ in xrange(K):
            bfs = [nei for node in bfs
                   for nei in neighbors[node]
                   if nei not in lookup]
            lookup |= set(bfs)
        return bfs

File: leetcode_java/src/main/java/LeetCodeJava/Tree/AllNodesDistanceKInBinaryTree.java
package LeetCodeJava.Tree;

// https://leetcode.com/problems/all-nodes-distance-k-in-binary-tree/description/

import LeetCodeJava.DataStructure.TreeNode;

import java.util.*;

/**
 * 863. All Nodes Distance K in Binary Tree
 * Solved
 * Medium
 * Topics
 * premium lock icon
 * Companies
 * Given the root of a binary tree, the value of a target node target, and an integer k, return an array of the values of all nodes that have a distance k from the target node.
 *
 * You can return the answer in any order.
 *
 *
 *
 * Example 1:
 *
 *
 * Input: root = [3,5,1,6,2,0,8,null,null,7,4], target = 5, k = 2
 * Output: [7,4,1]
 * Explanation: The nodes that are a distance 2 from the target node (with value 5) have values 7, 4, and 1.
 * Example 2:
 *
 * Input: root = [1], target = 1, k = 3
 * Output: []
 *
 *
 * Constraints:
 *
 * The number of nodes in the tree is in the range [1, 500].
 * 0 <= Node.val <= 500
 * All the values Node.val are unique.
 * target is the value of one of the nodes in the tree.
 * 0 <= k <= 1000
 *
 *
 */
public class AllNodesDistanceKInBinaryTree {

    // V0
//    public List<Integer> distanceK(TreeNode root, TreeNode target, int k) {
//
//    }

    // V0-1
    // IDEA: DFS + `Parent Map` + BFS (fixed by gpt)
    /**  IDEA:
     *
     * Why this works ?
     *
     *  Tree -> Graph -> BFS (visiting)
     *
     * 	•	From target you need to explore `all directions` reachable in k steps:
     * 	   ` left, right, and up (to parent). `
     * 	    Converting the tree to an `undirected graph` (children + parent edges)
     * 	    and then running BFS from target to depth k returns the desired nodes.
     *
     * 	•	visited ensures we don’t revisit nodes (which would otherwise make the BFS
     *    	incorrect / infinite once parent edges are present).
     *
     */
    List<Integer> res = new ArrayList<>();
    /**
     * parentMap stores `parent` pointers for every node (node -> parent).
     */
    Map<TreeNode, TreeNode> parentMap = new HashMap<>();

    public List<Integer> distanceK_0_1(TreeNode root, TreeNode target, int k) {
        if (root == null)
            return res;

        // 1️⃣ Build parent map for all nodes
        buildParentMap(root, null);

        /**
         * 	•	We do a breadth-first search (BFS) starting at the target node.
         * 	•	queue holds the frontier for BFS.
         * 	•	visited prevents revisiting nodes
         *   	(prevents cycles when we traverse parent links too).
         * 	•	Enqueue target and mark visited immediately so we don’t re-add it.
         * 	•	dist tracks current BFS depth (distance from target).
         */
        // 2️⃣ BFS starting from target, stop at distance k
        Queue<TreeNode> queue = new LinkedList<>();
        Set<TreeNode> visited = new HashSet<>();
        queue.offer(target);
        visited.add(target);
        int dist = 0;

        /**
         * 	-	Each loop iteration processes one BFS “level”
         * 	    (all nodes at the same distance from target).
         *
         * 	-	If current distance dist equals k, the nodes currently
         * 	    in queue are exactly the nodes at distance k.
         *
         *  - We collect their values and break out — no need to explore further.
         */
        while (!queue.isEmpty()) {
            int size = queue.size();
            if (dist == k) {
                // collect all nodes currently in the queue
                for (TreeNode node : queue) {
                    res.add(node.val);
                }
                break;
            }
            /**
             *  NOTE !!!
             *
             *   for each node, we visit `cur.left, cur.right, and its parent` via BFS
             *
             *
             * 	- Process the size nodes of the current level:
             * 	   - For each cur, try to move to `cur.left, cur.right, and its parent.`
             * 	   - visited.add(node) returns true only if node was not already present in visited.
             * 	     That both checks and marks in one call, so if (node != null && visited.add(node))
             * 	     queue.offer(node); is a compact idiom to avoid duplicates.
             *
             * 	-	After processing the whole level, increment dist and continue to next level.
             * 	- 	If BFS finishes without dist == k (e.g. k larger than tree diameter), res remains empty.
             * 	- 	Finally return res.
             */
            for (int i = 0; i < size; i++) {
                TreeNode cur = queue.poll();
                // explore neighbors: left, right, parent
                if (cur.left != null && visited.add(cur.left)) {
                    queue.offer(cur.left);
                }
                if (cur.right != null && visited.add(cur.right)) {
                    queue.offer(cur.right);
                }
                TreeNode parent = parentMap.get(cur);
                if (parent != null && visited.add(parent)) {
                    queue.offer(parent);
                }
            }
            dist++;
        }

        return res;
    }

    /**  NOTE !!! help func
     *
     * 	•	We need to be able to move `upwards` from any node (to parent).
     * 	A binary tree node only knows left/right children, so we precompute parents by a DFS.
     *
     *
     * 	•	Simple DFS that records parent of each node (parentMap.put(node, parent)).
     * 	•	For root we pass parent = null.
     * 	•	After this every node maps to its parent (or null for root).
     */
    private void buildParentMap(TreeNode node, TreeNode parent) {
        if (node == null)
            return;
        parentMap.put(node, parent);
        buildParentMap(node.left, node);
        buildParentMap(node.right, node);
    }

    // VO-2
    // IDEA: PURE DFS (gpt)
    List<Integer> res_0_2 = new ArrayList<>();

    public List<Integer> distanceK_0_2(TreeNode root, TreeNode target, int k) {
        dfs(root, target, k);
        return res_0_2;
    }

    // dfs() returns the distance from current node to target, or -1 if target is not in this subtree
    private int dfs(TreeNode node, TreeNode target, int k) {
        if (node == null)
            return -1;

        // 🎯 Base case: found target
        if (node == target) {
            // collect all nodes downward at distance k from target
            collectDownward(node, 0, k);
            return 0; // distance from target to itself
        }

        // 🔹 Search left subtree
        int leftDist = dfs(node.left, target, k);
        if (leftDist != -1) {
            int distFromNode = leftDist + 1;
            if (distFromNode == k) {
                res_0_2.add(node.val);
            } else {
                // find nodes in right subtree that are k - distFromNode - 0 away
                collectDownward(node.right, distFromNode + 1, k);
            }
            return distFromNode;
        }

        // 🔸 Search right subtree
        int rightDist = dfs(node.right, target, k);
        if (rightDist != -1) {
            int distFromNode = rightDist + 1;
            if (distFromNode == k) {
                res_0_2.add(node.val);
            } else {
                // find nodes in left subtree that are k - distFromNode away
                collectDownward(node.left, distFromNode + 1, k);
            }
            return distFromNode;
        }

        // target not found in either subtree
        return -1;
    }

    // helper: collect all nodes downward at distance k
    private void collectDownward(TreeNode node, int dist, int k) {
        if (node == null)
            return;
        if (dist == k) {
            res_0_2.add(node.val);
            return;
        }
        collectDownward(node.left, dist + 1, k);
        collectDownward(node.right, dist + 1, k);
    }

    // V1
    // https://leetcode.ca/2018-04-11-863-All-Nodes-Distance-K-in-Binary-Tree/
    // IDEA: DFS + HASHMAP
    private Map<TreeNode, TreeNode> p;
    private Set<Integer> vis;
    private List<Integer> ans;

    public List<Integer> distanceK_1(TreeNode root, TreeNode target, int k) {
        p = new HashMap<>();
        vis = new HashSet<>();
        ans = new ArrayList<>();
        parents(root, null);
        dfs(target, k);
        return ans;
    }

    private void parents(TreeNode root, TreeNode prev) {
        if (root == null) {
            return;
        }
        p.put(root, prev);
        parents(root.left, root);
        parents(root.right, root);
    }

    private void dfs(TreeNode root, int k) {
        if (root == null || vis.contains(root.val)) {
            return;
        }
        vis.add(root.val);
        if (k == 0) {
            ans.add(root.val);
            return;
        }
        dfs(root.left, k - 1);
        dfs(root.right, k - 1);
        dfs(p.get(root), k - 1);
    }


    // V2
    // IDEA: BFS
    // https://leetcode.com/problems/all-nodes-distance-k-in-binary-tree/solutions/3747835/two-bfs-c-java-python-beginner-friendly-jwora/
    public List<Integer> distanceK_2(TreeNode root, TreeNode target, int k) {
        List<Integer> ans = new ArrayList<>();
        Map<Integer, TreeNode> parent = new HashMap<>();
        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);

        while (!queue.isEmpty()) {
            int size = queue.size();
            for (int i = 0; i < size; i++) {
                TreeNode top = queue.poll();

                if (top.left != null) {
                    parent.put(top.left.val, top);
                    queue.offer(top.left);
                }

                if (top.right != null) {
                    parent.put(top.right.val, top);
                    queue.offer(top.right);
                }
            }
        }

        Map<Integer, Integer> visited = new HashMap<>();
        queue.offer(target);
        while (k > 0 && !queue.isEmpty()) {
            int size = queue.size();

            for (int i = 0; i < size; i++) {
                TreeNode top = queue.poll();

                visited.put(top.val, 1);

                if (top.left != null && !visited.containsKey(top.left.val)) {
                    queue.offer(top.left);
                }

                if (top.right != null && !visited.containsKey(top.right.val)) {
                    queue.offer(top.right);
                }

                if (parent.containsKey(top.val) && !visited.containsKey(parent.get(top.val).val)) {
                    queue.offer(parent.get(top.val));
                }
            }

            k--;
        }

        while (!queue.isEmpty()) {
            ans.add(queue.poll().val);
        }
        return ans;
    }


    // V3
    // IDEA: BFS
    // https://leetcode.com/problems/all-nodes-distance-k-in-binary-tree/solutions/7205194/the-best-detailed-explanation-let-you-ln-i610/
    public List<Integer> distanceK_3(TreeNode root, TreeNode target, int k) {

        HashMap<TreeNode, TreeNode> map = new HashMap<>();
        getParent(root, map);

        Queue<TreeNode> q = new LinkedList<>();
        q.add(target);
        int lvl = 0;
        HashSet<TreeNode> set = new HashSet<>();
        set.add(target);
        while (!q.isEmpty()) {

            int size = q.size();
            if (lvl == k)
                break;
            for (int i = 0; i < size; i++) {
                TreeNode curr = q.remove();

                if (curr.left != null && !set.contains(curr.left)) {
                    q.add(curr.left);
                    set.add(curr.left);
                }

                if (curr.right != null && !set.contains(curr.right)) {
                    q.add(curr.right);
                    set.add(curr.right);
                }

                TreeNode parent = map.get(curr);
                if (!set.contains(parent) && parent != null) {
                    q.add(parent);
                    set.add(parent);
                }
            }
            lvl++;
        }

        List<Integer> list = new ArrayList<>();
        while (!q.isEmpty()) {
            TreeNode curr = q.remove();
            list.add(curr.val);
        }
        return list;
    }

    public void getParent(TreeNode root, HashMap<TreeNode, TreeNode> map) {

        Queue<TreeNode> q = new LinkedList<>();
        q.add(root);

        while (!q.isEmpty()) {
            int size = q.size();

            for (int i = 0; i < size; i++) {
                TreeNode curr = q.remove();

                if (curr.left != null) {
                    q.add(curr.left);
                    map.put(curr.left, curr);
                }

                if (curr.right != null) {
                    q.add(curr.right);
                    map.put(curr.right, curr);
                }
            }
        }
    }


}
